### PR TITLE
[web] Defer injection of platform views until needed.

### PR DESCRIPTION
**Pull request by ditman**
This PR defers the injection of the contents of a Platform View into the DOM of the page, until the Platform View is really needed by a renderer.

This effectively means that a `platformView` will be injected into the DOM the first time that its `slot` is injected into the Shadow DOM of the Flutter web app.

This makes passing a `(flutter)ViewId` parameter from the framework unnecessary, even in a multi-view app.

The only cases in which this change might be breaking is those where an app tries to locate the just-created Platform View by looking into the DOM from the [`onPlatformViewCreated` callback](https://api.flutter.dev/flutter/widgets/HtmlElementView/onPlatformViewCreated.html). In those cases, [a fix like this](https://github.com/flutter/packages/pull/5660) is needed (use the **only [documented way](https://api.flutter.dev/flutter/dart-ui_web/PlatformViewRegistry/getViewById.html)** to obtain the Platform View contents from its `viewId`)

## Issues

Fixes https://github.com/flutter/flutter/issues/137287
Closes https://github.com/flutter/flutter/pull/136548

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/lib/src/engine/platform_views/content_manager.dart [hunk 1]
```diff
@@ -65,6 +65,11 @@ class PlatformViewManager {
     return _contents.containsKey(viewId);
   }
 
+  /// Returns the pre-rendered contents of [viewId], to inject them into the DOM.
+  DomElement getContents(int viewId) {
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/platform_views/content_manager.dart:69]: For consistency with the method below, should we add an assert-with-a-message here?

**Review by mdebbar — COMMENTED** [review]
You also need to change `engine/platform_dispatcher.dart`. It grabs `platformViewMessageHandler` from the implicit view.

Maybe we should make the `PlatformViewMessageHandler` a singleton? (since creating and disposing a platform view has nothing to do with a FlutterView).

**Comment by mdebbar** [comment]
That said, I like the this approach of deferring the insertion of platform views!

**Comment by ditman** [comment]
PTAL @mdebbar, I think I updated everything we discussed in the scrum today!

File: lib/web_ui/lib/src/engine/html/platform_view.dart [hunk 2]
```diff
@@ -3,22 +3,29 @@
...
+import '../window.dart';
 import 'surface.dart';
 
 /// A surface containing a platform view, which is an HTML element.
 class PersistedPlatformView extends PersistedLeafSurface {
-  PersistedPlatformView(this.viewId, this.dx, this.dy, this.width, this.height);
+  PersistedPlatformView(this.platformViewId, this.dx, this.dy, this.width, this.height);
 
-  final int viewId;
+  final int platformViewId;
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/html/platform_view.dart:15]: Thanks for renaming these!

File: lib/web_ui/lib/src/engine/platform_views/message_handler.dart [hunk 3]
```diff
@@ -141,39 +146,11 @@ class PlatformViewMessageHandler {
           params: arguments['params'],
         );
         return;
+      // TODO(web): Send `arguments` as a Map for `dispose` too!
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/platform_views/message_handler.dart:149]: 👍

File: lib/web_ui/lib/src/engine/view_embedder/dom_manager.dart [hunk 4]
```diff
@@ -204,6 +205,28 @@ class DomManager {
...
+  /// slot contents from the old `platformViewsHost` to the new one, but that
+  /// will cause the platformView to reset its state (an iframe will re-render,
+  /// text selections will be lost, video playback interrupted, etc...)
+  ///
+  /// Try not to move platform views across views!
+  void injectPlatformView(int platformViewId) {
+    // For now, we don't need anything fancier. If needed, this can be converted
+    // to a PlatformViewStrategy class for each web-renderer backend?
+    final DomElement pv = PlatformViewManager.instance.getSlottedContent(platformViewId);
+    // If pv is a descendant of _platformViewsContainer -> noop
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/view_embedder/dom_manager.dart:224]:
```suggestion
    // If pv is already a descendant of platformViewsHost -> noop
```

**Review by mdebbar — APPROVED** [review]
Looks good!! Ship it!

**Comment by ditman** [comment]
I'll update the tests, and land this. I tested all the example apps we have that use platform views and all seem to work normally.

File: lib/web_ui/lib/src/engine/view_embedder/dom_manager.dart [hunk 5]
```diff
@@ -204,6 +205,33 @@ class DomManager {
...
+      final DomElement pv = PlatformViewManager.instance.getSlottedContent(platformViewId);
+      // If pv is already a descendant of platformViewsHost -> noop
+      if (pv.parent == platformViewsHost) {
+        return;
+      }
+      platformViewsHost.append(pv);
+    } catch (e) {
+      // Do not fail, many tests expect this to not crash!
+      domWindow.console.warn(e);
+    }
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/view_embedder/dom_manager.dart:233]: Not super thrilled about this, but it fixes everything... opinions @mdebbar/@yjbanov?

**Comment by ditman** [comment]
Gold tests are legitimately broken by this change.

**Comment by ditman** [comment]
All right, I'm confident that I've fixed the goldens. The problem was the `html` and `canvaskit` renderers not injecting the slotted contents at the right time/place.

**Comment by ditman** [comment]
PTAL people, I think this is somewhat ready to land no! LMK about the locations of the view.dom.injectPlatformView(id) calls, those should happen always that are needed, but as few times as possible!

File: lib/web_ui/lib/src/engine/view_embedder/dom_manager.dart (hunk 5 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/view_embedder/dom_manager.dart:233]: What exactly is throwing here? The only thing I can think of is the assertion in `getSlottedContent`, but I would expect that to work if the tests are setting everything up correctly.

**Review by harryterkelsen — APPROVED** [review]
LGTM

File: lib/web_ui/lib/src/engine/view_embedder/dom_manager.dart (hunk 5 above)
Comment (ditman) [reply lib/web_ui/lib/src/engine/view_embedder/dom_manager.dart:233]: It is exactly the `getSlottedContent` assertion (or an unexpected null value if we don't assert). There's a bunch of tests in the framework that are not setting everything up correctly indeed.

**Review by mdebbar — APPROVED** [review]

File: lib/web_ui/test/engine/platform_dispatcher/platform_dispatcher_test.dart [hunk 6]
```diff
@@ -224,6 +224,32 @@ void testMain() {
...
+      bool onMetricsChangedCalled = false;
+      dispatcher.onMetricsChanged = () { 
+        onMetricsChangedCalled = true;
+      };
+
+      expect(onMetricsChangedCalled, isFalse);
+
+      dispatcher.viewManager.disposeAndUnregisterView(view2.viewId);
+
+      expect(onMetricsChangedCalled, isTrue);
```
Comment (mdebbar) [inline lib/web_ui/test/engine/platform_dispatcher/platform_dispatcher_test.dart:251]:
```suggestion
      expect(onMetricsChangedCalled, isTrue);
      dispatcher.dispose();
```

**Review by mdebbar — APPROVED** [review]
LGTM with a minor suggestion for the test.

File: lib/web_ui/test/engine/platform_dispatcher/platform_dispatcher_test.dart (hunk 6 above)
Comment (ditman) [reply lib/web_ui/test/engine/platform_dispatcher/platform_dispatcher_test.dart]: Good call, added!

**Comment by ditman** [comment]
If CI is happy, we're all happy. Go on auto-submit!

**Pull request merged**